### ISSUE TITLE
MQTT based Transport

### ISSUE THREAD

**Issue by JanEggers** [aspnet/SignalR#1768]
I dont want to spam more in https://github.com/aspnet/KestrelHttpServer/issues/1980 so il elaborte here what i was triing to archive.

I want to attach Signalr to my MQTT Broker, the Broker handles the basics of MQTT like QoS, Retain, etc...

But additionally in case a message with a certain topic comes in i need to do stuff. Currently i have my own message dispatcher but i think it would be more robust if i could use the signalr hub programming model instead of a switch (topic).

so @davidfowl suggested that i need to implement a custom IHubProtocol. i think it is great if i dont have to twiddle with the actual transport, but i thought the source of say JsonHubProtocol is a http connection. so if i want to do mqtt the transport needs to be a raw tcp server.

**Comment by anurse** [comment]
You'd need a few things to work together to make this happen:

1. You need a transport that MQTT can work over. Right now, we support Server-Sent Events, Long Polling and WebSockets. Of those three, only WebSockets is even remotely likely to work and I don't know if MQTT clients support that by default. So you'd have to build a TCP transport of some kind. I know @davidfowl has played with the idea but we have no plans for shipping it in RTM
2. You'd need a custom `IHubProtocol` that can decode MQTT messages and create SignalR invocations, and vice-versa.
3. We'd need a way to disable or modify the handshake process we go through. Before an `IHubProtocol` is instantiated, the client sends a "Handshake" payload indicating which protocol they wish to use. This handshake is always JSON. We have talked about a way to disable it though, since there are scenarios where the client and server already agree on which protocol to use.

**Comment by JanEggers** [comment]
> You need a transport that MQTT can work over.

that is why i posted in the bedrock thread that i miss the raw tcp transport @davidfowl posted as "2.2 and beyond" because I think that would be the ideal building block.

> Of those three, only WebSockets is even remotely likely to work

websockets work perfectly with mqtt i even build a little library on top of an mqtt library to use it with aspnetcore hosting. but i prefere to also support raw tcp.

https://github.com/chkr1011/MQTTnet/tree/master/Frameworks/MQTTnet.AspnetCore

> We'd need a way to disable or modify the handshake process we go through. 

i agree because that handshake is done with http so the handshake is http transport specific and just deals with payload negotiation (json or msgpack). for mqtt there is a connect message to specify mqtt version but that is totally unrelated.

**Comment by anurse** [comment]
> i agree because that handshake is done with http

Actually, it's the **negotiation** process that's done over HTTP (to the `/negotiate` endpoint). That's only needed in two cases: 1) You don't know which transports the server supports **or** 2) you want to connect with Long Polling or Server-Sent Events and require a connection ID be allocated before you can make requests (since there will be multiple requests related to the connection).

If the client knows the server supports WebSockets, it can connect the WebSocket directly, with no preceeding HTTP request (see https://github.com/aspnet/SignalR/blob/0285549aaaa6bf555fc67a589a461f06f8d0fe99/clients/ts/signalr/src/HttpConnection.ts#L78 where we skip negotiate if the client **explicitly** asks for WebSockets).

The "handshake" is a payload that is sent **over the WebSocket** to determine the _protocol_ to use (JSON or MsgPack).

**Comment by JanEggers** [comment]
i still havent found the code that glues together HttpConnectionContext and HubConnectionContext.

so if i would provide IHubProtocol and MQTTConnectionContext. how does signalr know when to use what? Or what do I have to do to make signalr aware of a MQTTConnectionContext? or do i need to provide a MqttConnectionManager similar to the HttpConnectionManager?

looking at HttpConnectionDispatcher: (i know its unrelated since all transports in there are http) why is the following information pulled out of statics. wouldnt it be better if HttpConnectionDispatcher get the available transports from di, so one could add another transport without touching the dispatcher?

https://github.com/aspnet/SignalR/blob/12b5e173b38a03ffb8a96c26f1c316a4e232e8fe/src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs#L27-L46

**Comment by davidfowl** [comment]
> i still havent found the code that glues together HttpConnectionContext and HubConnectionContext.

The ConnectionContext is the exchange type that ConnectionHandler use. The implementation of that for the http layer is the HttpConnectionContext. I don't think you need any of that. You're just parsing I believe.

> looking at HttpConnectionDispatcher: (i know its unrelated since all transports in there are http) why is the following information pulled out of statics. wouldnt it be better if HttpConnectionDispatcher get the available transports from di, so one could add another transport without touching the dispatcher?

We don't allow plugging in transports at this layer nor do we have plans to make it extensible.

**Comment by JanEggers** [comment]
I added a mqtt endpoint in my program like in the samples, that opened up the tcp. It broke in the Handshake as @anurse explained above.

```
 .UseKestrel(o => {
                    o.ListenAnyIP(1883, l => 
                             l.UseHub<MqttHub>());
                })
```
from my point of view I would need a special HubConnectionHandler that doesnt do a handshake but rather a generic version so i can specify that my MqttHubProtocol shall be used. like:

```
HubConnectionHandler<THub, TProtocol> : ConnectionHandler 
        where THub : Hub
        where TProtocol : IHubProtocol

and extension 

public static IConnectionBuilder UseHub<THub, TProtocol>(this IConnectionBuilder connectionBuilder) where THub : Hub
where TProtocol : IHubProtocol
        {
            return connectionBuilder.UseConnectionHandler<HubConnectionHandler<THub, TProtocol>>();
        }
```

I would like to provide a PR but i didnt settle with my head of department that im allowed to sign your CLA, so I would be happy if you could include something.

I also could not prototype further because this Property is internal:

https://github.com/aspnet/SignalR/blob/ef30e2e2df459ceeb18ccf84ab29ecb78efe14ec/src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs#L73

**Comment by JanEggers** [comment]
I overcome that internal issue with reflection so wI took it a little further and hit another issue:

MQTT Specifies a Connect Message with response I can parse it but how am I supposed to send the response? 

my MqttHubProtocol is called by some Dispatcher is there any way to write something to the output without beeing called?

**Comment by JanEggers** [comment]
is there any way to move all the code i have in there to the hubprotocol?

https://github.com/JanEggers/Playground/blob/SignalR/MQTT/Playground.core/Hubs/MqttHub.cs

I basicly need to respond to incoming calls but with the current api this seems to be only possible with returnvalues from the hub.

I would like to avoid having a constraint on the hub

**Comment by davidfowl** [comment]
You can't respond from the protocol layer, it's there just to parse data. You can respond from your custom ConnectionHandler though.

**Comment by JanEggers** [comment]
thx for the suggestion I moved the code from Hub to ConnectionHandler. now the hub is clean again. :)

only remaining issue is that this is internal but i have to set it. (worked around with reflection) https://github.com/aspnet/SignalR/blob/ef30e2e2df459ceeb18ccf84ab29ecb78efe14ec/src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs#L73

**Comment by davidfowl** [comment]
@JanEggers We'll fix that, there are other scenarios we have in mind that necessitate that.

**Comment by JanEggers** [comment]
@davidfowl I updated to rc1 and tried to build a client side implementation. but im currently blocked by https://github.com/aspnet/SignalR/blob/d5f62f7b5a2495995b6f53033c0e6fa7bf8bda17/src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs#L507

is it possible to involve IHubProtocol in that process so i can create a mqttconnect packet if HandshakeRequestMessage is used and you can do your thing in your protocols?

**Comment by davidfowl** [comment]
@JanEggers hmm, we don't have the same hooks on the client side as on the server but it looks like we'd need that. Seems like you'll have to hack and use a custom copy of the HubConnection to get around that. We can open up more in the next release. We're pretty much locking down 1.0.0 right now so no big features are happening at the moment. 

How are you using the client side in an MQTT application?

**Comment by JanEggers** [comment]
@davidfowl just to let you know: I postponed the Signar Mqtt adaption for msg dispatching.

For now I just use the bedrock transport story instead of System.Net.Socket and gained 30 times rps improvement so thx for the awesome api.

**Comment by JanEggers** [comment]
Hi there I encountered an issue when triing to use the connectionhandler for websockets.

it is able to receive packages but in the browser I get this Error after receiving some frames:

WebSocket connection to 'ws://localhost:5000/mqtt' failed: Could not decode a text frame as UTF-8.

![image](https://user-images.githubusercontent.com/6747376/42408901-66e86224-81d2-11e8-8f32-a2c265008355.png)

code is here:
https://github.com/JanEggers/MQTTnet/tree/ConnectionHandlerWithWebsockets

**Comment by JanEggers** [comment]
found it myself #2159
